File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        backtrack(0,nums,ans);
        return ans;
    }
    private void backtrack(int index,int[] nums,List<List<Integer>> ans){
        if(index==nums.length){
		// new list to store the data and copy to the ans 
            List<Integer> ds = new ArrayList<>();
            
            for(int i=0;i<nums.length;i++) ds.add(nums[i]);
            
            ans.add(ds);
            return;
        }
        for(int i=index;i<nums.length;i++){
            swap(i,index,nums);
            backtrack(index+1,nums,ans);
            swap(i,index,nums);
        }
    }
    private void swap(int i, int j,int[] nums){
        int t= nums[i];
        nums[i]=nums[j];
        nums[j]=t;
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void swap(int i, int j, vector<int>& nums) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
        return;
    }
    void backtrack(int i, vector<int>& nums, vector<vector<int>>& res) {
        if(i == nums.size()) {
            vector<int> temp;
            for(int num:nums) temp.push_back(num);
            res.push_back(temp);
            return;
        }
        else {
            for(int j=i; j<nums.size(); j++) {
                swap(i, j, nums);
                backtrack(i+1, nums, res);
                swap(i, j, nums);
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        backtrack(0, nums, res);
        return res;
    }
};